filebeat/input/filestream/copytruncate_prospector.go

```diff
@@ -0,0 +1,344 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package filestream
+
+import (
+	"os"
+	"path/filepath"
+	"regexp"
+	"sort"
+	"time"
+
+	"github.com/urso/sderr"
+
+	loginp "github.com/elastic/beats/v7/filebeat/input/filestream/internal/input-logfile"
+	input "github.com/elastic/beats/v7/filebeat/input/v2"
+	"github.com/elastic/beats/v7/libbeat/logp"
+	"github.com/elastic/go-concert/unison"
+)
+
+const (
+	copyTruncateProspectorDebugKey = "copy_truncate_file_prospector"
+	copiedFileIdx                  = 0
+)
+
+// sorter is required for ordering rotated log files
+type sorter interface {
+	sort([]rotatedFileInfo)
+}
+
+// rotatedFileInfo stores the file information of a rotated file.
+type rotatedFileInfo struct {
+	path string
+	src  loginp.Source
+	ts   time.Time
+}
+
+func (f rotatedFileInfo) String() string {
+	return f.path
+}
+
+// rotatedFilestream includes the information of the original file
+// and its identifier, and the rotated file.
+type rotatedFilestream struct {
+	originalSrc loginp.Source
+	rotated     []rotatedFileInfo
+}
+
+func newRotatedFilestreams(cfg *copyTruncateConfig) *rotatedFilestreams {
+	var sorter sorter
+	sorter = &defaultSorter{}
+	if cfg.DateFormat != "" {
+		sorter = &dateSorter{cfg.DateFormat}
+	}
+	return &rotatedFilestreams{
+		table:  make(map[string]*rotatedFilestream, 0),
+		count:  cfg.Count,
+		sorter: sorter,
+	}
+}
+
+// defaultSorter sorts rotated log files that have a numeric suffix.
+// Example: apache.log.1, apache.log.2
+type defaultSorter struct{}
+
+func (s *defaultSorter) sort(files []rotatedFileInfo) {
+	sort.Slice(
+		files,
+		func(i, j int) bool {
+			return files[i].path < files[j].path
+		},
+	)
+}
+
+// dateSorter sorts rotated log files that have a date suffix
+// based on the configured format.
+// Example: apache.log-21210526, apache.log-20210527
+type dateSorter struct {
+	format string
+}
+
+func (s *dateSorter) sort(files []rotatedFileInfo) {
+	sort.Slice(
+		files,
+		func(i, j int) bool {
+			return s.GetTs(&files[i]).After(s.GetTs(&files[j]))
```

Comment:
as the func is support to implement `less`, I would prefer to use `Before` and change the order of the arguments. Kinda makes it easier to reason.

In comparison to the `defaultSorter`, it looks like the ordering is opposite. The `defaultSorter` puts the newest files first, while the `dateSorter` puts the newest file at the end (basically implements `f[j] < f[i]`).


The `GetTS` will be called multiple times for a single file, depending on the sorting algorithm being used. Alternatively you might opt to cache the timestamp in the `rotatedFileInfo`.

Reply:
It is already cached.